### ISSUE TITLE
Most of CMS seems to work; except for this test case (detached, binary)

### ISSUE THREAD

**Issue by dirkx** [aaronhuggins/node-libas2#9]
OpenSSL code to sign _detached_ (normal and smime is fine)
```
echo Something to Sign > payload
openssl req -new -x509 -nodes -keyout x509.key -out x509.pub -subj /CN=CoronaPub
openssl cms -sign -signer x509.pub -inkey x509.key -outform DER -out signature-cms.bin -in payload
openssl cms -verify -CAfile x509.pub -inkey x509.pub -inform DER -in signature-cms.bin -content payload
```

 against:

```
const libas2 = require("libas2");
const ForgeVerify_1 = require("./node_modules/libas2/dist/src/AS2Crypto/ForgeVerify-debug.js");
const verify = ForgeVerify_1.verify;

const forge = require('node-forge');
const fs = require('fs');


var certAsPem = fs.readFileSync('x509.pub');
var cert = forge.pki.certificateFromPem(certAsPem)
var payloadAsBin = fs.readFileSync('payload');
var sig_as_der = fs.readFileSync('signature-cms.bin');

console.log("Parsing");
const asn1 = forge.asn1.fromDer(sig_as_der);

console.log("Decoding");
const msg = forge.pkcs7.messageFromAsn1(asn1);
console.log(msg);

console.log("Binding");
verify.bind(msg);

console.log("Verify");
const verified = verify({
        msg: msg,
        certificate: cert,
        detached: payloadAsBin
});

console.log("Result");
console.log(verified);
```

**Comment by ahuggins-nhs** [comment]
Hmm, I would have thought my tests covered this scenario. I'm working on getting this into a production app over the next two weeks and I'll try to address soon.

**Comment by dirkx** [comment]
Agreed - all else worked right away. This was the exception. 

I've included the generation script -- and happy to stand corrected !

**Comment by ahuggins-nhs** [comment]
Oh, wait... I think I see the issue. The Buffer you get from reading in a file needs to be converted to a `forge` buffer before being passed to verify:
```js
const forgeBuffer = forge.util.createBuffer(payloadAsBin)
```

The reason I would not implement in `libas2` is that I think that's how things are handled in the API for `forge` if you look over their API. The `ForgeVerify.ts` module will eventually be submitted as a JS commit back to the Forge project, and I would like to keep parity with their patterns to make it easier to give back to their community. If you can think of a good reason to code an explicit dependency in `ForgeVerify.ts` on node class `Buffer`, let me know and I can look at that, but I believe the above code snippet will resolve your issue.

**Comment by ahuggins-nhs** [comment]
I might still need to add a test for your openssl example, so I'll leave this open and circle back unless you close the issue, @dirkx .

**Comment by dirkx** [comment]
No luck yet.

```
const libas2 = require("libas2");
const ForgeVerify_1 = require("./node_modules/libas2/dist/src/AS2Crypto/ForgeVerify-wth-debug.js");
const verify = ForgeVerify_1.verify;

const forge = require('node-forge');
const fs = require('fs');
 
 
var certAsPem = fs.readFileSync('x509.pub');
var cert = forge.pki.certificateFromPem(certAsPem)
var payloadAsBin = fs.readFileSync('payload');
var sig_as_der = fs.readFileSync('signature-cms.bin');

console.log("Parsing");
const asn1 = forge.asn1.fromDer(forge.util.createBuffer(sig_as_der));

console.log("Decoding");
const msg = forge.pkcs7.messageFromAsn1(asn1);
console.log(msg);

console.log("Binding");
verify2 = verify.bind(msg);

console.log("Verify");
console.log(msg);

const verified = verify2({ 
        msg: msg,
        certificate: cert, 
        detached: forge.util.createBuffer(payloadAsBin)
});

console.log("Result");
console.log(verified);
```

**Comment by ahuggins-nhs** [comment]
Adding bug label back. Will run your code samples this afternoon.

**Comment by ahuggins-nhs** [comment]
@dirkx, what is your environment (OS, version of node, version of openssl, etc.), and can you provide the error message you're receiving?

**Comment by dirkx** [comment]
OSX10.15.5 or FreeBSD 12.1; nodejs 13.8.0_0 or v12.16.1; the error is a 'false' on the final output (shown below)

    signature: '£üfÄ\u0010+\tÜëøùÈAfvÝþÉ\u0003"l\u0016Ç¯\u0017@\u0018>0,ãàò~çd0\u0004\u001bÔÆ\u0011Ì(¦ í\u001c\u0018öæsP\u0004çd¥Q±>^\u000b¦|g·\u0006K¤}\'ÐÑnâØe>³\u0014÷}k¼Î\u0003íìÔÖ\u0010ðÊ=\u0011îr"¥n\u0013bÊ°¥·²²èã7¡·À`ñ ©ãñFç^³Â~±`§yµ´\u001cÀ\u0019<¹¼møä§\u0004\u001c1²ºs\u0011æslÅµ5$\tò\u001aaVà\u001cD«´º©?\u000fcaÄßÎ8\u001eu\rô\\v9¼n\u0005Ü\u0013LÿÁôÖ\b¹nsPé¥à=\u0003ý?/ý= \u0003Û6b\u0014Ê¤E',
    signerInfos: [ [Object] ]
  }
}
Result
false

**Comment by dirkx** [comment]
actual script, testcases
[Archive.zip](https://github.com/ahuggins-nhs/node-libas2/files/4827629/Archive.zip)

**Comment by ahuggins-nhs** [comment]
@dirkx, I actually did a ton of debugging yesterday and I'm working on swapping out the library. There were more than a few additional cases I came up with based on what you gave me where node-forge was failing. So I took a harder look at node-forge.

Node-forge doesn't fully implement CMS versions past RFC 2315 (at least from code comments in their repo) and did not have `verify` implemented at all (which is why I rolled my own attempt). PKI.js, on top of their node webcrypto polyfill `@peculiar/webcrypto`, fully implements more recent CMS RFCs with more parity to OpenSSL than node-forge. After a bit more testing, I think it will be better to have a more complete crypto foundation to meet the needs of the AS2 protocol, especially if I don't need to roll a custom verify routine.

In short, @dirkx, thank you for finding this issue.

**Comment by ahuggins-nhs** [comment]
I've reached out to the PKI.js community through an issue. A lot of them spend far more time steeped in cryptography than I do, and I'm at a complete loss at this point after having replaced the signing/verification library. I'm still able to reproduce this issue, so I've added it as a permanent test case to the mocha suite so I don't lose sight of this issue.

**Comment by ahuggins-nhs** [comment]
@dirkx, [please see the resolution on the PKIjs issue](https://github.com/PeculiarVentures/PKI.js/issues/279#issuecomment-650344854).

Based on findings there, this is an OpenSSL issue and not a PKIjs or AS2 issue. The AS2 mime-handling already implements canonicalizing per RFC 5751, I was very careful to make that a part of the library and comply with the RFC. If you are going to use pure implementations of CMS like PKIjs or node-forge (which I'm replacing anyway), you'll need to canonicalize any detached payload before verifying it against a signature produced by OpenSSL.
